File: algos/solutions/a2fastslowpointers/PY/circulararray/algo.py
def circular_array_loop(nums):
    size = len(nums)

    for i in range(size):
        slow = fast = i
        forward = nums[i] > 0

        while True:
            slow = next_step(slow, nums[slow], size)
            if is_not_cycle(nums, forward, slow):
                break

            fast = next_step(fast, nums[fast], size)
            if is_not_cycle(nums, forward, fast):
                break

            fast = next_step(fast, nums[fast], size)
            if is_not_cycle(nums, forward, fast):
                break

            if slow == fast:
                return True

    return False


def next_step(pointer, value, size):
    return (pointer + value) % size


def is_not_cycle(nums, prev_direction, pointer):
    curr_direction = nums[pointer] > 0

    # If direction changes, cycle is invalid
    if prev_direction != curr_direction:
        return True

    # A self-loop is not a valid cycle
    if next_step(pointer, nums[pointer], len(nums)) == pointer:
        return True

    return False

File: algos/solutions/a2fastslowpointers/JS/circulararray/algo.js
// circulararray/algo.js

function circularArrayLoop(nums) {
    const n = nums.length;

    // Helper function to calculate the next index
    const nextIndex = (current) => {
        return ((current + nums[current]) % n + n) % n; // Ensure positive modulo
    };

    for (let i = 0; i < n; i++) {
        if (nums[i] === 0) continue; // Skip already visited elements

        let slow = i, fast = i;
        const direction = nums[i] > 0; // Determine the direction of the loop

        while (true) {
            slow = nextIndex(slow);
            if (nums[slow] === 0 || (nums[slow] > 0) !== direction) break;

            fast = nextIndex(fast);
            if (nums[fast] === 0 || (nums[fast] > 0) !== direction) break;

            fast = nextIndex(fast);
            if (nums[fast] === 0 || (nums[fast] > 0) !== direction) break;

            if (slow === fast) {
                if (slow === nextIndex(slow)) break; // Single-element loop
                return true;
            }
        }

        // Mark all elements in the current path as visited
        let current = i;
        while (nums[current] !== 0) {
            const next = nextIndex(current);
            nums[current] = 0;
            current = next;
        }
    }

    return false;
}

module.exports = circularArrayLoop;
